arches/app/permissions/arches_standard.py

```diff
@@ -0,0 +1,925 @@
+            and "users_without_read_perm" in search_result["_source"]["permissions"]
+            and (
+                user.id
+                in search_result["_source"]["permissions"]["users_without_read_perm"]
+            )
+        ) and user_can_read
+
+        user_can_edit = len(self.get_editable_resource_types(user)) > 0
+        result["can_edit"] = (
+            "permissions" in search_result["_source"]
+            and "users_without_edit_perm" in search_result["_source"]["permissions"]
+            and (
+                user.id
+                in search_result["_source"]["permissions"]["users_without_edit_perm"]
+            )
+            and user_can_edit
+        )
+        result["is_principal"] = (
+            "permissions" in search_result["_source"]
+            and "principal_user" in search_result["_source"]["permissions"]
+            and user.id in search_result["_source"]["permissions"]["principal_user"]
+        )
+        return result
+
+
+class PermissionBackend(ObjectPermissionBackend):  # type: ignore
+    def has_perm(self, user_obj: User, perm: str, obj: Model | None = None) -> bool:
+        # check if user_obj and object are supported (pulled directly from guardian)
+        support, user_obj = check_support(user_obj, obj)
+        if not support:
+            return False
+
+        if "." in perm:
+            app_label, perm = perm.split(".")
+            if obj is None:
+                raise ValueError("Passed perm has app label of '%s' and obj is None")
+            if app_label != obj._meta.app_label:
+                raise WrongAppError(
+                    "Passed perm has app label of '%s' and "
+                    "given obj has '%s'" % (app_label, obj._meta.app_label)
+                )
+
+        obj_checker: ObjectPermissionChecker = CachedObjectPermissionChecker(
+            user_obj, obj
+        )
+        explicitly_defined_perms = obj_checker.get_perms(obj)
+
+        if len(explicitly_defined_perms) > 0:
+            if "no_access_to_nodegroup" in explicitly_defined_perms:
+                return False
+            else:
+                return perm in explicitly_defined_perms
+        else:
+            user_checker = CachedUserPermissionChecker(user_obj)
+            return user_checker.user_has_permission(perm)
+
+
+class CachedUserPermissionChecker:
+    """
+    A permission checker that leverages the 'user_permission' cache to check user-level user permissions.
+    """
+
+    def __init__(self, user: User):
+        user_permission_cache = caches["user_permission"]
+        current_user_cached_permissions = user_permission_cache.get(str(user.pk), {})
+
+        if current_user_cached_permissions.get("user_permissions"):
+            user_permissions = current_user_cached_permissions.get("user_permissions")
+        else:
+            user_permissions = set()
+            user.groups.prefetch_related("permissions")
+
+            for group in user.groups.all():
```

Comment:
I was thinking like:
```suggestion
            for group in user.groups.prefetch_related("permissions").all():
```